luacompiler/src/lib/irgen/register_map.rs

```diff
@@ -1,43 +1,80 @@
-use std::{cell::RefCell, collections::HashMap};
+use std::{collections::HashMap, vec::Vec};
 
-/// Represents a structure that is used to keep track of the mapping
-/// between Lua variables and register ids.
-pub struct RegisterMap {
-    reg_count: RefCell<u8>,
-    reg_map: RefCell<HashMap<String, u8>>,
+/// Represents a tuple which is used to specify the lifetime of a register.
+/// For example if a register is first used by the 4th instruction of the bytecode, and
+/// used last by the 7th instruction, the register's lifetime would be (4, 8).
+pub struct Lifetime(usize, usize);
+
+impl Lifetime {
+    pub fn new(sp: usize) -> Lifetime {
+        Lifetime(sp, sp + 1)
+    }
+
+    /// Get the start point of the register.
+    pub fn start_point(&self) -> usize {
+        self.0
+    }
+
+    /// Get the end point of the register.
+    pub fn end_point(&self) -> usize {
+        self.1
+    }
+
+    fn set_end_point(&mut self, ep: usize) {
+        self.1 = ep
+    }
+}
+
+/// Represents a structure that is used to map Lua variables to registers, and to keep
```

Comment:
Is there one register map per function or scope or file or ...?

Reply:
Currently it is per file. Should I update the documentation to make it more clear?

Reply:
How are you going to move from per-file to per-function? Shouldn't you (at least minimally) think about that now before things get baked in?

Reply:
I was thinking that each `compile_function` method will create an instance of `RegMap`. I will have a global one, for global variables, and then one for each function.

Reply:
How will this work with nested scopes? Lua has (at least) two scopes (function and module) (and maybe others I'm not aware of).

Reply:
Then an even better solution, each variable is named `var_<scopenum>`. Then I can keep using only one map, which makes things easier I guess.

```lua
x = 2 -- will be named 'x' or 'x_mod'
if x > 2 then
  local x  = 2 -- will be named 'x_scope1'
else
  local x = 3 -- will be named 'x_scope2'
```

Reply:
That is highly unlikely to work. I leave that as an exercise to you.